functions/go/set-labels/transformer/labels.go

```diff
@@ -0,0 +1,168 @@
+package transformer
+
+import (
+	"fmt"
+	"github.com/GoogleContainerTools/kpt-functions-sdk/go/fn"
+	"sigs.k8s.io/yaml"
+	"strings"
+)
+
+// the similar struct esxits in resid.GVK, but there is no function to create an GVK struct without using kyaml
+type FieldSpec struct {
+	Group              string `json:"group,omitempty" yaml:"group,omitempty"`
+	Version            string `json:"version,omitempty" yaml:"version,omitempty"`
+	Kind               string `json:"kind,omitempty" yaml:"kind,omitempty"`
+	Path               string `json:"path,omitempty" yaml:"path,omitempty"` // seperated by /
+	CreateIfNotPresent bool   `json:"create,omitempty" yaml:"create,omitempty"`
+}
+
+func SetLabels(rl *fn.ResourceList) (bool, error) {
+	transformer := LabelTransformer{}
+	if err := transformer.Config(rl.FunctionConfig); err != nil {
+		rl.Results = append(rl.Results, fn.ErrorResult(err))
+		return false, nil
+	}
+	if err := transformer.Transform(rl.Items); err != nil {
+		rl.Results = append(rl.Results, fn.ErrorResult(err))
+		return false, nil
+	}
+	return true, nil
+}
+
+type LabelTransformer struct {
+	// Desired labels
+	NewLabels map[string]string `json:"labels,omitempty" yaml:"labels,omitempty"`
+	// FieldSpecs is deprecated, please use AdditionalLabelFields instead.
+	FieldSpecs []FieldSpec `json:"fieldSpecs,omitempty" yaml:"fieldSpecs,omitempty"`
+	// AdditionalLabelFields is used to specify additional fields to add labels.
+	AdditionalLabelFields []FieldSpec `json:"additionalLabelFields,omitempty" yaml:"additionalLabelFields,omitempty"`
+	// Results is used to track labels that have been applied
+	Results fn.Results
+}
+
+// Config parse the functionConfig kubeObject to the fields in the LabelTransformer
+func (p *LabelTransformer) Config(o *fn.KubeObject) error {
+	// parse labels to NewLabels
+	switch {
+	case o.IsEmpty():
+		return fmt.Errorf("FunctionConfig is missing. Expect `ConfigMap` or `SetNamespace`")
+	case o.IsGVK("", "", "ConfigMap"):
+		p.NewLabels = o.NestedStringMapOrDie("data")
+		if len(p.NewLabels) == 0 {
+			return fmt.Errorf("`data` should not be empty")
+		}
+	case o.IsGVK(fnConfigGroup, fnConfigAPIVersion, legacyFnConfigKind):
+		fallthrough
+	case o.IsGVK(fnConfigGroup, fnConfigAPIVersion, fnConfigKind):
+		p.NewLabels = o.NestedStringMapOrDie("labels")
+		if len(p.NewLabels) == 0 {
+			return fmt.Errorf("failed to configure function: input label list cannot be empty")
+		}
+	default:
+		return fmt.Errorf("unknown functionConfig Kind=%v ApiVersion=%v, expect `%v` or `ConfigMap`",
+			o.GetKind(), o.GetAPIVersion(), fnConfigKind)
+	}
+	// add default fields
+	if err := p.addDefaultLabelFields(); err != nil {
+		return err
+	}
+	// add additional fields
+	if o.IsGVK(fnConfigGroup, fnConfigAPIVersion, fnConfigKind) {
+		arr, exist, err := o.NestedSlice("additionalLabelFields")
+		if err != nil {
+			return fmt.Errorf("`additionalLabelFields` format is wrong")
```

Comment:
We should surface `err` as well.

Reply:
OK. I changed the implementation to use o.Get() and surface the error